src/Integration.Vsix.UnitTests/ErrorList/SonarErrorListDataSourceTests.cs

```diff
@@ -127,69 +129,169 @@ public void AddFactory_ExistingSinkManagersAreNotifiedOfNewFactory()
+            testSubject.Subscribe(mockSink2.Object);
 
             var mockFactory = new Mock<ITableEntriesSnapshotFactory>();
 
             testSubject.AddFactory(mockFactory.Object);
 
-            CheckFactoryWasAdded(mockManager1, mockFactory.Object);
-            CheckFactoryWasAdded(mockManager2, mockFactory.Object);
+            CheckFactoryWasAdded(mockSink1, mockFactory.Object);
+            CheckFactoryWasAdded(mockSink2, mockFactory.Object);
         }
 
         [TestMethod]
-        public void RemoveSinkManager_ManagerIsNoLongerNotified()
+        public void RemoveSink_IsNoLongerNotified()
         {
             var testSubject = new SonarErrorListDataSource(mockTableManagerProvider.Object);
 
-            var mockSinkManager = new Mock<ISinkManager>();
-            testSubject.AddSinkManager(mockSinkManager.Object);
+            var mockSink = new Mock<ITableDataSink>();
+            var disposeToken = testSubject.Subscribe(mockSink.Object);
 
             // 1. Manager is registered -> should be notified
             testSubject.RefreshErrorList();
             testSubject.RefreshErrorList();
 
-            mockSinkManager.Verify(x => x.UpdateSink(), Times.Exactly(2));
+            mockSink.Verify(x => x.FactorySnapshotChanged(null), Times.Exactly(2));
 
-            // 2. Unregister manager -> no longer notified
-            mockSinkManager.Reset();
-            testSubject.RemoveSinkManager(mockSinkManager.Object);
+            // 2. Unsubscribe -> no longer notified
+            mockSink.Reset();
+            disposeToken.Dispose();
             testSubject.RefreshErrorList();
 
-            mockSinkManager.Verify(x => x.UpdateSink(), Times.Never);
+            mockSink.Verify(x => x.FactorySnapshotChanged(It.IsAny<ITableEntriesSnapshotFactory>()), Times.Never);
         }
 
         [TestMethod]
         public void RemoveFactory_ExistingSinkManagersAreNotified()
         {
             var testSubject = new SonarErrorListDataSource(mockTableManagerProvider.Object);
 
-            var mockManager1 = new Mock<ISinkManager>();
-            var mockManager2 = new Mock<ISinkManager>();
-            testSubject.AddSinkManager(mockManager1.Object);
-            testSubject.AddSinkManager(mockManager2.Object);
+            var mockSink1 = new Mock<ITableDataSink>();
+            var mockSink2 = new Mock<ITableDataSink>();
+            testSubject.Subscribe(mockSink1.Object);
+            testSubject.Subscribe(mockSink2.Object);
 
             var mockFactory = new Mock<ITableEntriesSnapshotFactory>();
 
             testSubject.RemoveFactory(mockFactory.Object);
 
-            CheckFactoryWasRemoved(mockManager1, mockFactory.Object);
-            CheckFactoryWasRemoved(mockManager2, mockFactory.Object);
+            CheckFactoryWasRemoved(mockSink1, mockFactory.Object);
+            CheckFactoryWasRemoved(mockSink2, mockFactory.Object);
+        }
+
+        #region Exception handling tests
```

Comment:
The changes above this line are just fix-up so the tests work against `ITableDataSink` rather than `ISinkManager`.

The tests below this line are the exception handling tests copied from the `SinkManagerTests` and cleaned up slightly.  There are tests for each of `AddFactory`, `RemoveFactory` and `RefreshErrorList` to check that critical errors are not suppressed but non-critical errors are.